### ISSUE TITLE
Run VM apps at low process priority

### ISSUE THREAD

**Issue by davidpanderson** [BOINC/boinc#2785]
Currently VM apps (i.e. vbox_headless.exe) run at normal process priority.
This slows down system response; they should run at low priority.

I recall that VBox has no provision for running VMs at low priority.
If this is still the case, perhaps the client can identify VM processes
and lower the priority itself.

**Comment by lfield** [comment]
Duplicate of #2273 but proposes a different solution.

**Comment by davidpanderson** [comment]
Goals:
1) stagger the start of VM apps at client startup
	currently they can saturate I/O when several start at once
2) lower the priority of VM apps.
	currently they run at normal process priority

Challenge:
There is AFAIK no way to find the process IDs of the VMs the client starts,
so that we can distinguish them from other VMs

Proposal:

1) Staggering:
a) wait 30 sec before starting any VM app
b) wait 15 sec between starting VM apps

2) lowering priority:

Add to ACTIVE_TASK
	double start_time: when started
	bool demoted: VM job; has been demoted

When we enumerate processes (procinfo_setup())
make a list of the processes running VBoxheadless.exe.
Do this at startup, before we start any jobs.

Subsequently, we enumerate all processes every 10 sec.
for each VM process P not in previous VM list
	look for an active VM task T that
		- was started after the previous process scan
		- has not been demoted;
	find the earliest such.
	lower its priority; flag as demoted

This isn't foolproof; it may demote non-BOINC VMs or fail to demote BOINC VMs.

**Comment by JuhaSointusalo** [comment]
> 1. stagger the start of VM apps at client startup
>    currently they can saturate I/O when several start at once

There is an older ticket open for staggered startup of tasks. #1293

> 2. lower the priority of VM apps.
>     currently they run at normal process priority

VirtualBox locks down the VM process pretty tight. If I try to change the priority of the process that is really running the VM all I get back is Access denied. It would seem to me that the only really working way to control the priority of VM tasks is implementing the control in VirtualBox as #2273 says.

Since it's not possible to implement both of the goals of this ticket at the same time (different code bases and all that) I would suggest that this ticket is closed and David copies the proposed implementation of staggered startup to #1293. This is so that

1. we don't have duplicate issues
2. the tickets can be closed as soon as they are done instead of having our backlog full of half completed tickets

**Comment by davidpanderson** [comment]
I forgot about that- yes, the vboxwrapper gets the process ID from the log.
So now the issue is how to lower its priority.
From what Laurence says, VBox currently makes that impossible by limiting access rights
to the process.
So maybe the solution here is for one of us to fix VBox - either by adding
a priority arg when creating the VM (COM or cmdline),
or by restoring that access right.
I can't imagine that being able to change the priority of a VM is a security issue.

**Comment by AenBleidd** [comment]
With the release of [BUDA (Docker)](https://github.com/BOINC/boinc/wiki/BUDA-overview) that is quite stable now and already used by [BOINC Central](https://boinc.berkeley.edu/central/) and [LHC@home](https://lhcathome.cern.ch/lhcathome/index.php) we stop an active maintenance of the VBox-based applications.
Support of the VBox-based applications will not be dropped, but only the critical issues will be fixed without any new features implementation.
Thus this ticket is closed without any further actions.
